docs/source/core/core.rst

```diff
@@ -630,19 +631,248 @@ To auto-resume training set the following via YAML or CLI:
+        # the axes layout is different
+        assert btc_void.compare(btc_spct_bad) == NeuralTypeComparisonResult.INCOMPATIBLE
+        # Note that even though VoidType is compatible with every other type, other types are not compatible with VoidType!
+        # It is one-way compatibility
+        assert btc_spctr.compare(btc_void) == NeuralTypeComparisonResult.INCOMPATIBLE
+
+Element type inheritance
+~~~~~~~~~~~~~~~~~~~~~~~~
+Neural types in NeMo support Python inheritance between element types.
+Consider an example where you want to develop a Neural Module which performs data augmentation for all kinds of spectrograms.
+In ASR two types of spectrograms are frequently used: mel and mfcc. To express this we will create two 3 classes to express
+element's types: `SpectrogramType`, `MelSpectrogramType(SpectrogramType)`, `MFCCSpectrogramType(SpectrogramType)`.
+
+.. code-block:: python
+
+        input = NeuralType(('B', 'D', 'T'), SpectrogramType())
+        out1 = NeuralType(('B', 'D', 'T'), MelSpectrogramType())
+        out2 = NeuralType(('B', 'D', 'T'), MFCCSpectrogramType())
+
+        # MelSpectrogram and MFCCSpectrogram are not interchangeable.
+        assert out1.compare(out2) == NeuralTypeComparisonResult.INCOMPATIBLE
+        assert out2.compare(out1) == NeuralTypeComparisonResult.INCOMPATIBLE
+        # Type comparison detects that MFCC/MelSpectrogramType is a kind of SpectrogramType and can be accepted.
+        assert input.compare(out1) == NeuralTypeComparisonResult.GREATER
+        assert input.compare(out2) == NeuralTypeComparisonResult.GREATER
+
+Custom element types
+~~~~~~~~~~~~~~~~~~~~
+It is possible to create user-defined element types to express the semantics of elements in your tensors.
+To do so, you need to inherit and implement abstract methods of `nemo.core.neural_types.elements.ElementType` class
+
+.. autoclass:: nemo.core.neural_types.elements.ElementType
+
+Note that element types can be parametrized. Consider this example where it distinguishes between audio sampled at 8Khz and 16Khz.
+
+.. code-block:: python
+
+    audio16K = NeuralType(axes=('B', 'T'), elements_type=AudioSignal(16000))
+    audio8K = NeuralType(axes=('B', 'T'), elements_type=AudioSignal(8000))
+
+    assert audio8K.compare(audio16K) == NeuralTypeComparisonResult.SAME_TYPE_INCOMPATIBLE_PARAMS
+    assert audio16K.compare(audio8K) == NeuralTypeComparisonResult.SAME_TYPE_INCOMPATIBLE_PARAMS
+
+Enforcing dimensions
+~~~~~~~~~~~~~~~~~~~~
+In addition to specifying tensor layout and elements' semantics, neural types also allow you to enforce tensor dimensions.
+You will have to use long notation to specify dimensions. Short notation only allows you to specify axes semantics and assumes
+arbitrary dimensions.
+
+.. code-block:: python
+
+        type1 = NeuralType(
+        (AxisType(AxisKind.Batch, 64), AxisType(AxisKind.Time, 10), AxisType(AxisKind.Dimension, 128)),
```

Comment:
Do we enforce this inside .compare()? I know `typecheck()` does not do actual shape enforcement. Ill add to my todos if .compare() doesnt do it currently.

Reply:
yes

Reply:
I actually copied unittests :D